src/resdk/collection_tables.py

```diff
@@ -0,0 +1,290 @@
+
+        self.cache_dir = cache_dir
+        if self.cache_dir is None:
+            self.cache_dir = default_cache_dir()
+        if not os.path.exists(self.cache_dir):
+            os.makedirs(self.cache_dir)
+
+        self._table_samples_ = []  # type: List[TableSample]
+        self._meta = None  # type: Optional[pd.DataFrame]
+        self._expressions = {}  # type: Dict[str, pd.DataFrame]
+        self._gene_map_ = {}  # type: Dict[str, str]
+        self._modified = None  # type: Optional[datetime]
+
+    @lru_cache(maxsize=16)
+    def expressions(self, preprocess=None) -> pd.DataFrame:
+        """Return TPM data in a table format.
+
+        If a data preprocess is set then its executed and its results get cached in memory.
+
+        :param preprocess: a callable that preprocess expression data
+        :return: table of TPM
+        """
+        cache = load_cache(self.cache_dir, self.collection.slug, TPM, self.modified)
+        if cache is not None:
+            self._expressions[TPM] = cache
+
+        if TPM not in self._expressions:
+            self._expressions[TPM] = self._download_expressions(TPM)
+
+        save_cache(self._expressions[TPM],
+                   self.cache_dir, self.collection.slug, TPM, self.modified)
+
+        _expressions = self._expressions[TPM]
+        if preprocess is not None:
+            _expressions = preprocess(_expressions.copy())
+        return _expressions
+
+    def counts(self) -> pd.DataFrame:
+        """Return counts in a table format.
+
+        :return: table of counts
+        """
+        cache = load_cache(self.cache_dir, self.collection.slug, COUNTS, self.modified)
+        if cache is not None:
+            self._expressions[COUNTS] = cache
+
+        if COUNTS not in self._expressions:
+            self._expressions[COUNTS] = self._download_expressions(COUNTS)
+
+        save_cache(self._expressions[COUNTS],
+                   self.cache_dir, self.collection.slug, COUNTS, self.modified)
+
+        return self._expressions[COUNTS]
+
+    def metadata(self) -> pd.DataFrame:
+        """Return samples metadata in a table format.
+
+        :return: table of metadata
+        """
+        cache = load_cache(self.cache_dir, self.collection.slug, META, self.modified)
+        if cache is not None:
+            self._meta = cache
+
+        if self._meta is None:
+            self._meta = self._get_metadata()
+
+        save_cache(self._meta,
+                   self.cache_dir, self.collection.slug, META, self.modified)
+
+        return self._meta
+
+    @property
+    def _table_samples(self) -> List[TableSample]:
+        """Return a list of TableSample objects containing expression files urls and metadata."""
+        if not self._table_samples_:
+            self._table_samples_ = self._get_table_samples()
+        return self._table_samples_
+
+    @property
+    def modified(self) -> datetime:
```

Comment:
Any particular logic for how methods are ordered? Readers of this module would benefit if relevant methods were kept close, e.g. `modifed` and `_get_modified` are adjacent. Consider reordering methods.